File: demos/jans-chip/android/app/src/main/java/io/jans/chip/repository/LoginResponseRepository.java
package io.jans.chip.repository;

import android.content.Context;
import android.util.Log;

import java.util.List;
import java.util.UUID;

import io.jans.chip.AppDatabase;
import io.jans.chip.modal.LoginResponse;
import io.jans.chip.modal.OIDCClient;
import io.jans.chip.modal.OPConfiguration;
import io.jans.chip.modal.OperationError;
import io.jans.chip.retrofit.RetrofitClient;
import io.jans.chip.modal.SingleLiveEvent;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class LoginResponseRepository {
    public static final String TAG = "LoginRepository";
    private final SingleLiveEvent<LoginResponse> loginResponseLiveData = new SingleLiveEvent<>();
    Context context;
    AppDatabase appDatabase;

    private LoginResponseRepository(Context context) {
        this.context = context;
        appDatabase = AppDatabase.getInstance(context);
    }

    private static LoginResponseRepository dcrRepository;

    public static LoginResponseRepository getInstance(Context context) {
        if (dcrRepository == null) {
            dcrRepository = new LoginResponseRepository(context);
        }
        return dcrRepository;
    }

    public SingleLiveEvent<LoginResponse> processlogin(String usernameText, String passwordText) {
        // Get OPConfiguration and OIDCClient
        List<OPConfiguration> opConfigurationList = appDatabase.opConfigurationDao().getAll();
        if (opConfigurationList == null || opConfigurationList.isEmpty()) {
            loginResponseLiveData.setValue(setErrorInLiveObject("OpenID configuration not found in database."));
            return loginResponseLiveData;
        }
        List<OIDCClient> oidcClientList = appDatabase.oidcClientDao().getAll();
        if (oidcClientList == null || oidcClientList.isEmpty()) {
            loginResponseLiveData.setValue(setErrorInLiveObject("OpenID client not found in database."));
            return loginResponseLiveData;
        }
        OPConfiguration opConfiguration = opConfigurationList.get(0);
        OIDCClient oidcClient = oidcClientList.get(0);
        Log.d(TAG,"Authorization Challenge Endpoint :: "+ opConfiguration.getAuthorizationChallengeEndpoint());
        // Create a call to request an authorization challenge
        Call<LoginResponse> call = RetrofitClient.getInstance(opConfiguration.getIssuer()).getAPIInterface().
                getAuthorizationChallenge(oidcClient.getClientId(),
                        usernameText,
                        passwordText,
                        UUID.randomUUID().toString(),
                        UUID.randomUUID().toString(),
                        true,
                        opConfiguration.getAuthorizationChallengeEndpoint());
        call.enqueue(new Callback<LoginResponse>() {
            @Override
            public void onResponse(Call<LoginResponse> call, Response<LoginResponse> response) {
                if (response.code() == 200) {
                    LoginResponse responseFromAPI = response.body();
                    Log.d(TAG, "processlogin Response :: getAuthorizationCode ::" + responseFromAPI.getAuthorizationCode());

                    if (responseFromAPI.getAuthorizationCode() != null && !responseFromAPI.getAuthorizationCode().isEmpty()) {
                        responseFromAPI.setSuccessful(true);
                        loginResponseLiveData.setValue(responseFromAPI);
                    } else {
                        loginResponseLiveData.setValue(setErrorInLiveObject("Error in fetching OP Configuration. Authorization code is empty"));
                    }
                } else {
                    loginResponseLiveData.setValue(setErrorInLiveObject("Error in  generating authorization code.\n Error code: " + response.code() + "\n Error message: " + response.message()));
                }
            }

            @Override
            public void onFailure(Call<LoginResponse> call, Throwable t) {
                Log.e(TAG, "Inside processlogin :: onFailure :: " + t.getMessage());
                loginResponseLiveData.setValue(setErrorInLiveObject("Inside processlogin :: onFailure :: " + t.getMessage()));
            }
        });
        return loginResponseLiveData;
    }

    private LoginResponse setErrorInLiveObject(String errorMessage) {
        OperationError operationError = new OperationError.Builder()
                .title("Error")
                .message(errorMessage)
                .build();
        LoginResponse loginResponse = new LoginResponse(false, operationError);
        return loginResponse;
    }
}

File: demos/jans-chip/android/app/src/main/java/io/jans/chip/repository/LoginResponseRepository.kt
package io.jans.chip.repository

import android.content.Context
import android.util.Log
import io.jans.chip.AppDatabase
import io.jans.chip.model.LoginResponse
import io.jans.chip.model.OPConfiguration
import io.jans.chip.model.UserInfoResponse
import io.jans.chip.retrofit.ApiAdapter
import retrofit2.Response
import java.util.UUID

class LoginResponseRepository(context: Context) {
    private val TAG = "LoginResponseRepository"
    private val appDatabase = AppDatabase.getInstance(context);
    var obtainedContext: Context = context
    private var loginResponse: LoginResponse? = LoginResponse(null)
    suspend fun processLogin(
        usernameText: String,
        passwordText: String?,
        authMethod: String,
        assertionResultRequest: String?
    ): LoginResponse? {
        // Get OPConfiguration and OIDCClient
        try {
            val opConfigurationList: List<OPConfiguration>? =
                appDatabase.opConfigurationDao().getAll()
            if (opConfigurationList == null || opConfigurationList.isEmpty()) {
                loginResponse?.isSuccessful = false
                loginResponse?.errorMessage = "OpenID configuration not found in database."
                return loginResponse
            }
            val oidcClientList = appDatabase.oidcClientDao().getAll()
            if (oidcClientList == null || oidcClientList.isEmpty()) {
                loginResponse?.isSuccessful = false
                loginResponse?.errorMessage = "OpenID client not found in database."
                return loginResponse
            }
            val opConfiguration: OPConfiguration = opConfigurationList[0]
            val oidcClient = oidcClientList[0]
            Log.d(
                TAG,
                "Authorization Challenge Endpoint :: " + opConfiguration.authorizationChallengeEndpoint
            )
            // Create a call to request an authorization challenge
            val response: Response<LoginResponse> =
                ApiAdapter.getInstance(opConfiguration.issuer).getAuthorizationChallenge(
                    oidcClient.clientId,
                    usernameText,
                    passwordText,
                    UUID.randomUUID().toString(),
                    UUID.randomUUID().toString(),
                    true,
                    "passkey",
                    authMethod,
                    assertionResultRequest,
                    opConfiguration.authorizationChallengeEndpoint
                )

            if (response.code() != 200) {
                loginResponse?.isSuccessful = false
                loginResponse?.errorMessage =
                    "Error in login. Check Username/ Password."
                Log.d(
                    TAG,
                    "Error in login. Check Username/ Password. ErrorCode : ${response.code()} and ErrorMessage: ${response.message()}"
                )
                return loginResponse
            }
            loginResponse = response.body()
            if (!response.isSuccessful || loginResponse == null) {
                loginResponse?.isSuccessful = false
                loginResponse?.errorMessage =
                    "Error in login. ErrorCode : ${response.code()} and ErrorMessage: ${response.message()}"
                return loginResponse
            }

            Log.d(
                TAG,
                "processlogin Response :: getAuthorizationCode :: ${loginResponse?.authorizationCode}"
            )
            loginResponse?.isSuccessful = true
            return loginResponse
        } catch (e: Exception) {
            Log.e(TAG, "Error in  login. ${e.message}".trimIndent())
            loginResponse?.isSuccessful = false
            loginResponse?.errorMessage =
                "Error in login. Error message: ${e.message}"
            return loginResponse
        }
    }

    suspend fun isAuthenticated(accessToken: String?): Boolean {
        var userInfoResponseRepository: UserInfoResponseRepository =
            UserInfoResponseRepository(obtainedContext)
        var userInfoResponse: UserInfoResponse? =
            userInfoResponseRepository.getUserInfo(accessToken)
        return userInfoResponse?.let { it -> it?.isSuccessful } ?: false
    }

    suspend fun getUserInfoWithAccessToken(accessToken: String?): UserInfoResponse? {
        var userInfoResponseRepository: UserInfoResponseRepository =
            UserInfoResponseRepository(obtainedContext)
        return userInfoResponseRepository.getUserInfo(accessToken)

    }
}